File: src/main/java/algorithms/search/no200/no_200.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        row = len(grid)
        col = len(grid[0])
        visited = [[False] * col for _ in range(row)]
        ans = 0
        for i in range(row):
            for j in range(col):
                if grid[i][j] == '1' and visited[i][j] == False:
                    ans += 1
                    self.search(grid, i, j, visited)
        return ans

    def search(self, grid: List[List[str]], i: int, j: int, visited: List[List[bool]]) -> None:
        if i < 0 or i == len(grid) or j < 0 or j == len(grid[0]): return
        if grid[i][j] == '0': return
        if visited[i][j]: return
        visited[i][j] = True
        self.search(grid, i + 1, j, visited)
        self.search(grid, i - 1, j, visited)
        self.search(grid, i, j + 1, visited)
        self.search(grid, i, j - 1, visited)

File: src/main/java/algorithms/search/no200/No200.java
package algorithms.search.no200;

/**
 * 200. 岛屿数量
 * @author Sean Yu
 * @since 2024/5/22 13:53
 */

public class No200 {
}

class Solution200 {
    boolean[][] visited;

    public int numIslands(char[][] grid) {
        visited = new boolean[grid.length][grid[0].length];
        int ans = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    ans++;
                    search(grid, i, j);
                }
            }
        }
        return ans;
    }

    private void search(char[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i == grid.length || j == grid[0].length) return;
        if (grid[i][j] == '0') return;
        //now grid[i][j] == 1
        if (visited[i][j]) return;
        visited[i][j] = true;
        search(grid, i - 1, j);
        search(grid, i + 1, j);
        search(grid, i, j - 1);
        search(grid, i, j + 1);
    }
}
